lib/puppet/type/file/target.rb

```diff
@@ -63,8 +63,13 @@ def mklink
     def insync?(currentvalue)
       if [:nochange, :notlink].include?(self.should) or @resource.recurse?
         return true
-      elsif ! @resource.replace? and Puppet::FileSystem.exist?(@resource[:path])
-        return true
+      elsif ! @resource.replace?
+        begin
+          Puppet::FileSystem.lstat(@resource[:path])
+          return true
+        rescue Errno::ENOENT
+          return false
+        end
```

Comment:
I mentioned this in the ticket so I'll put a link here. I think we need to take into account whether we're managing the link or following the link: https://tickets.puppetlabs.com/browse/PUP-10214?focusedCommentId=710253&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-710253

Reply:
Thanks @joshcooper; I also responded to that point in the ticket:
https://tickets.puppetlabs.com/browse/PUP-10214?focusedCommentId=807542&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-807542

In short, I think that the `links` parameter is for something different, and my test results are consistent with that.

I can also explain by referring to the documentation:
https://puppet.com/docs/puppet/7/types/file.html#links

> How to handle links during file actions. During file copying, follow will copy the target file instead of the link and manage will copy the link itself. When not copying, manage will manage the link, and follow will manage the file to which the link points.

Creating a symlink via `ensure => 'link'` does not do any file copying. There's no source file to copy--the symlink is created directly from the value of the `target` parameter.

`links => 'manage` is, instead, for when using the `source` parameter, which specifies:
> A source file, which will be copied into place on the local system.
[...]
> Unlike content, the source attribute can be used to recursively copy directories if the recurse attribute is set to true or remote. If a source directory contains symlinks, use the links attribute to specify whether to recreate links or follow them.

I'm maybe 95% sure of this, as a long-time user, though not as an author of the code.


As far as I can tell, my change is scoped to only apply only to the `target` parameter, since it is in `lib/puppet/type/file/target.rb`.

-Corey

Reply:
@bugfood Coming back to this after awhile. The `links` parameter is used during catalog application, to tell the agent whether it's managing the link or the thing the link points to. For example, when setting the owner for a `file` resource we have to call `chown` or `lchown` https://github.com/puppetlabs/puppet/blob/1a53bf7d66275ad4b4928484bb1314dc32e84677/lib/puppet/provider/file/posix.rb#L87-L91

When a `file` resource has a `source => 'puppet:///modules/...` parameter (instead of `content`), then the agent requests file metadata from puppetserver and passes along `links` as an HTTP request parameter https://github.com/puppetlabs/puppet/blob/1a53bf7d66275ad4b4928484bb1314dc32e84677/lib/puppet/http/service/file_server.rb#L36-L37 

Puppetserver uses that information to generate metadata, for example, permissions, file checksums, etc. It uses the supplied `links` parameter to decide whether to follow links that might be present in an environment on disk. Puppetserver returns the metadata and the agent merges that with the metadata specified explicitly in the catalog (with the latter taking precedence). The agent then uses the metadata to determine if its local file is insync or not.

Now back to this issue, puppet behaves correctly with dangling symlinks, `replace => false` and `links => manage`:

```
❯ rm -f /tmp/foo && rm -f /tmp/bar && echo orig > /tmp/bar && ln -s bar /tmp/foo && rm /tmp/bar
❯ file /tmp/foo  
/tmp/foo: broken symbolic link to bar
❯ file /tmp/bar
/tmp/bar: cannot open `/tmp/bar' (No such file or directory)
❯ cat links.pp 
file { '/tmp/foo':
  ensure  => link,
  target => 'bar',
  replace => false,
  links   => manage
}
❯ bundle exec puppet apply links.pp                                                                     
Notice: Compiled catalog for localhost in environment production in 0.01 seconds
Notice: Applied catalog in 0.02 seconds
```

Changing `links => follow` triggers one of the variations in https://puppet.atlassian.net/browse/PUP-5830. Specifically, puppet attempts to follow the link, sees that the thing the link points to doesn't exist, and so tries to create the link, but fails because it already exists:

```
❯ rm -f /tmp/foo && rm -f /tmp/bar && echo orig > /tmp/bar && ln -s bar /tmp/foo && rm /tmp/bar
❯ cat links.pp                                                                                 
file { '/tmp/foo':
  ensure  => link,
  target => 'bar',
  replace => false,
  links   => follow
}
❯ bundle exec puppet apply links.pp --trace                                                             
Notice: Compiled catalog for localhost in environment production in 0.01 seconds
Error: Could not set 'link' on ensure: File exists @ syserr_fail2_in - /tmp/foo (file: /home/josh/work/puppet/links.pp, line: 1)
/home/josh/.rbenv/versions/3.2.2/lib/ruby/3.2.0/fileutils.rb:722:in `symlink'
/home/josh/.rbenv/versions/3.2.2/lib/ruby/3.2.0/fileutils.rb:722:in `block in ln_s'
/home/josh/.rbenv/versions/3.2.2/lib/ruby/3.2.0/fileutils.rb:2497:in `fu_each_src_dest0'
/home/josh/.rbenv/versions/3.2.2/lib/ruby/3.2.0/fileutils.rb:720:in `ln_s'
/home/josh/work/puppet/lib/puppet/file_system/file_impl.rb:130:in `symlink'
/home/josh/work/puppet/lib/puppet/file_system.rb:262:in `symlink'
/home/josh/work/puppet/lib/puppet/type/file/target.rb:55:in `block (2 levels) in mklink'
/home/josh/work/puppet/lib/puppet/util/suidmanager.rb:65:in `asuser'
/home/josh/work/puppet/lib/puppet/type/file/target.rb:48:in `block in mklink'
/home/josh/work/puppet/lib/puppet/type/file/target.rb:47:in `chdir'
/home/josh/work/puppet/lib/puppet/type/file/target.rb:47:in `mklink'
/home/josh/work/puppet/lib/puppet/type/file/ensure.rb:105:in `block (2 levels) in <module:Puppet>'
/home/josh/work/puppet/lib/puppet/property.rb:489:in `set'
/home/josh/work/puppet/lib/puppet/property.rb:569:in `sync'
/home/josh/work/puppet/lib/puppet/type/file/ensure.rb:191:in `sync'
```

Given the fact that `links => manage` works correctly and is the default, it doesn't appear this PR is needed. Are you able to reproduce the original issue using the current HEAD of `main`?

Reply:
@joshcooper thanks for coming back to look at this. I refreshed my memory and looked at the code.

I see what you mean about the `links` parameter being used during catalog application, at least for setting ownership as you pointed out. I think I wasn't aware of that before. Thanks for the reference to `lib/puppet/provider/file/posix.rb`.

It doesn't look like `lib/puppet/type/file/target.rb` has any check for `:manage` like that, though.

I can still reproduce the issue with current `main` as of today. In your test case above, you're already setting the symlink target to the value specified in the puppet code.
* `ln -s bar /tmp/foo`
* `target => 'bar',`

Here's an adapted test case to show the issue:
```
$ rm -f /tmp/foo /tmp/bar && ln -s /tmp/bar /tmp/foo
$ file /tmp/foo /tmp/bar
/tmp/foo: broken symbolic link to /tmp/bar
/tmp/bar: cannot open `/tmp/bar' (No such file or directory)
$ cat > links.pp <<EOF
> file { '/tmp/foo':
  ensure  => 'link',
  target  => 'baz',
  replace => false,
  links   => 'manage',
}
EOF
$ RUBYLIB=./lib bin/puppet apply links.pp
Notice: Compiled catalog for corey.home.arpa in environment production in 0.01 seconds
Notice: /Stage[main]/Main/File[/tmp/foo]/target: target changed '/tmp/bar' to 'baz'
Notice: Applied catalog in 0.02 seconds
$ file /tmp/foo /tmp/bar
/tmp/foo: broken symbolic link to baz
/tmp/bar: cannot open `/tmp/bar' (No such file or directory)
```

I'm running puppet directly from the source dir, but I think it should be the same if you use `bundle exec`.

I can try putting in a conditional as per your original comment:
>  I think we need to take into account whether we're managing the link or following the link

-Corey

Reply:
@joshcooper I pushed an update to the PR that makes my change only apply to `links => 'manage'`. I added some testing of this as well. Please let me know what you think of the current PR.

Side note:

The more I think about this, I really don't know what puppet could be expected to do with `ensure => 'link', links => 'follow'`. This particular combination of parameters seems invalid to me and should perhaps be disallowed.
* If puppet uses OS resolution of symlinks (`stat`), then the file being managed will never be seen as a symlink. Dangling symlinks will resolve as file-not-found and valid symlinks will resolve as the target file. Puppet can't really follow the symlink becuse it doesn't know where the symlink points, or even if the file is a symlink.
* If puppet uses manual resolution of symlinks (`lstat` and `readlink`), then it will need to do so recursively in order to match OS behavior. At this point, which file does puppet actually manage? The top-level symlink, or the last symlink in a chain, or the file pointed to by the last symlink? None of these seem like they would be expected behaviors.

-Corey